File: validator/validate.js
#!/usr/bin/env node
/**
 * SMCU Protocol — Standalone Validator (JavaScript)
 * Usage: node validate.js <path_to_entry.json>
 *        node validate.js examples/software-dev.json
 *        node validate.js examples/software-dev.json --json
 */

const fs   = require("fs");
const path = require("path");

const REQUIRED_FIELDS    = ["id","taxonomie","contenu","preuve","validation","confiance","statut","source","visibilite","portee","cycle_de_vie"];
const REQUIRED_TAXONOMIE = ["domaine","sous_domaine","macro","micro"];
const REQUIRED_CONTENU   = ["type","gravite","description","solution"];

const VALID_STATUTS    = ["actif","conteste","revoque","obsolete","archive","en_attente"];
const VALID_GRAVITES   = ["critique","majeur","mineur"];
const VALID_TYPES      = ["erreur","regle","avertissement","bonne_pratique"];
const VALID_VISIBILITE = ["public","organisation","prive"];
const VALID_PREUVES    = ["test_unitaire","test_integration","cas_clinique","simulation","peer_review","audit_securite"];
const VALID_DECISIONS  = ["approuve","rejete","ameliorer"];
const VALID_AFFICHAGE  = ["public","pseudonyme","anonyme"];

// ─────────────────────────────────────────────
function validate(entry) {
  const errors   = [];
  const warnings = [];

  const check = (condition, msg) => { if (!condition) errors.push(msg); };

  // Required top-level fields
  for (const f of REQUIRED_FIELDS) {
    check(f in entry, `Champ requis manquant : '${f}'`);
  }

  // ID format
  if ("id" in entry) {
    const parts = entry.id.split("-");
    check(parts.length >= 3, `ID '${entry.id}' invalide. Format attendu : DOM-TYPE-HASH`);
  }

  // Taxonomie
  if ("taxonomie" in entry) {
    const t = entry.taxonomie;
    for (const f of REQUIRED_TAXONOMIE) {
      check(f in t && t[f], `taxonomie.${f} requis et non vide`);
    }
    if (!t.tags || t.tags.length === 0) {
      warnings.push("taxonomie.tags vide — réduit la découvrabilité");
    }
  }

  // Contenu
  if ("contenu" in entry) {
    const c = entry.contenu;
    for (const f of REQUIRED_CONTENU) {
      check(f in c && c[f], `contenu.${f} requis et non vide`);
    }
    if ("gravite" in c) check(VALID_GRAVITES.includes(c.gravite), `contenu.gravite invalide : '${c.gravite}'`);
    if ("type"    in c) check(VALID_TYPES.includes(c.type),       `contenu.type invalide : '${c.type}'`);
    if (!("exemple" in c)) warnings.push("contenu.exemple absent — recommandé pour adoption");
  }

  // Preuve
  if ("preuve" in entry) {
    const p = entry.preuve;
    check("type" in p, "preuve.type requis");
    if ("type" in p) check(VALID_PREUVES.includes(p.type), `preuve.type invalide : '${p.type}'`);
    if (!p.reference) warnings.push("preuve.reference absent — réduira la confiance des validateurs");
  }

  // Validation / votes
  if ("validation" in entry) {
    const v = entry.validation;
    if ("votes_requis" in v) {
      check(Number.isInteger(v.votes_requis) && v.votes_requis >= 3,
            `validation.votes_requis doit être ≥ 3 (actuel : ${v.votes_requis})`);
    }
    if ("votes" in v) {
      v.votes.forEach((vote, i) => {
        check("votant_did" in vote, `vote[${i}].votant_did manquant`);
        check("decision" in vote && VALID_DECISIONS.includes(vote.decision),
              `vote[${i}].decision invalide. Acceptés : ${VALID_DECISIONS.join(", ")}`);
        check("timestamp" in vote, `vote[${i}].timestamp manquant`);
      });
    }
    if ("votes_positifs" in v && "votes_total" in v) {
      check(v.votes_positifs <= v.votes_total, "votes_positifs ne peut pas dépasser votes_total");
    }
  }

  // Confiance
  if ("confiance" in entry && "score" in entry.confiance) {
    const s = entry.confiance.score;
    check(s >= 0 && s <= 100, `confiance.score doit être entre 0 et 100 (actuel : ${s})`);
    if (s < 40) warnings.push(`confiance.score faible (${s}) — non appliquée automatiquement`);
  }

  // Statut
  if ("statut" in entry) {
    check(VALID_STATUTS.includes(entry.statut), `statut invalide : '${entry.statut}'`);
  }

  // Source
  if ("source" in entry) {
    const s = entry.source;
    if (s.affichage) {
      check(VALID_AFFICHAGE.includes(s.affichage), `source.affichage invalide : '${s.affichage}'`);
      if (s.affichage === "pseudonyme") check(s.pseudonyme, "source.pseudonyme requis");
      if (s.affichage === "anonyme")    check(s.preuve_zkp,  "source.preuve_zkp requis");
      if (["public","pseudonyme"].includes(s.affichage)) check(s.did, "source.did requis");
    }
  }

  // Visibilité
  if ("visibilite" in entry) {
    const niv = entry.visibilite.niveau;
    check(VALID_VISIBILITE.includes(niv), `visibilite.niveau invalide : '${niv}'`);
    if (niv === "organisation" && !entry.visibilite.organisation_id) {
      errors.push("visibilite.organisation_id requis quand niveau='organisation'");
    }
  }

  // Relations
  if ("relations" in entry && !("parent" in entry.relations)) {
    warnings.push("relations.parent absent — hiérarchie recommandée");
  }

  return {
    valid:    errors.length === 0,
    entry_id: entry.id || "inconnu",
    errors,
    warnings,
    summary:  `${errors.length} erreur(s), ${warnings.length} avertissement(s)`
  };
}

// ─────────────────────────────────────────────
function printReport(result) {
  const R = "\x1b[0m", BOLD = "\x1b[1m";
  const G = "\x1b[92m", RED = "\x1b[91m", Y = "\x1b[93m";

  const status = result.valid ? `${G}✅ VALIDE${R}` : `${RED}❌ INVALIDE${R}`;
  console.log(`\n${BOLD}SMCU Validator — ${result.entry_id}${R}`);
  console.log(`Status : ${status}`);
  console.log(`Résumé : ${result.summary}\n`);

  if (result.errors.length)   { console.log(`${RED}${BOLD}Erreurs:${R}`);          result.errors.forEach(e => console.log(`  ${RED}✗${R} ${e}`)); }
  if (result.warnings.length) { console.log(`\n${Y}${BOLD}Avertissements:${R}`);   result.warnings.forEach(w => console.log(`  ${Y}⚠${R} ${w}`)); }
  console.log();
}

// ─────────────────────────────────────────────
const args     = process.argv.slice(2);
const filePath = args.find(a => !a.startsWith("--"));
const jsonMode = args.includes("--json");

if (!filePath) {
  console.log("Usage: node validate.js <chemin_vers_entree.json> [--json]");
  process.exit(1);
}

if (!fs.existsSync(filePath)) {
  console.error(`Erreur : fichier introuvable : '${filePath}'`);
  process.exit(1);
}

let entry;
try {
  entry = JSON.parse(fs.readFileSync(filePath, "utf-8"));
} catch (e) {
  console.error(`Erreur JSON : ${e.message}`);
  process.exit(1);
}

const result = validate(entry);
printReport(result);
if (jsonMode) console.log(JSON.stringify(result, null, 2));

process.exit(result.valid ? 0 : 1);

File: validator/validate.py
#!/usr/bin/env python3
"""
SMCU Protocol — Standalone Validator CLI
Usage: python validate.py <path_to_entry.json>
       python validate.py examples/software-dev.json
"""

import sys
import json
import math
import os

# ─────────────────────────────────────────────
REQUIRED_FIELDS = [
    "id", "taxonomie", "contenu", "preuve",
    "validation", "confiance", "statut",
    "source", "visibilite", "portee", "cycle_de_vie"
]
REQUIRED_TAXONOMIE  = ["domaine", "sous_domaine", "macro", "micro"]
REQUIRED_CONTENU    = ["type", "gravite", "description", "solution"]
REQUIRED_VALIDATION = ["votes_requis", "votes", "votes_positifs", "votes_total"]

VALID_STATUTS    = ["actif", "conteste", "revoque", "obsolete", "archive", "en_attente"]
VALID_GRAVITES   = ["critique", "majeur", "mineur"]
VALID_TYPES      = ["erreur", "regle", "avertissement", "bonne_pratique"]
VALID_VISIBILITE = ["public", "organisation", "prive"]
VALID_PREUVES    = [
    "test_unitaire", "test_integration", "cas_clinique",
    "simulation", "peer_review", "audit_securite"
]
VALID_DECISIONS  = ["approuve", "rejete", "ameliorer"]
VALID_AFFICHAGE  = ["public", "pseudonyme", "anonyme"]

CONFIDENCE_THRESHOLD_HIGH   = 80
CONFIDENCE_THRESHOLD_MEDIUM = 60
CONFIDENCE_THRESHOLD_LOW    = 40


def _check(condition, errors, message):
    if not condition:
        errors.append(message)


def validate(entry: dict) -> dict:
    errors   = []
    warnings = []

    # ── Top-level required fields
    for f in REQUIRED_FIELDS:
        _check(f in entry, errors, f"Champ requis manquant : '{f}'")

    # ── ID format
    if "id" in entry:
        parts = entry["id"].split("-")
        _check(len(parts) >= 3, errors,
               f"ID '{entry['id']}' invalide. Format attendu : DOM-TYPE-HASH (ex: DEV-ERR-A3F2)")

    # ── Taxonomie
    if "taxonomie" in entry:
        t = entry["taxonomie"]
        for f in REQUIRED_TAXONOMIE:
            _check(f in t and t[f], errors, f"taxonomie.{f} requis et non vide")
        if "tags" not in t or not t.get("tags"):
            warnings.append("taxonomie.tags vide — réduit la découvrabilité de l'entrée")

    # ── Contenu
    if "contenu" in entry:
        c = entry["contenu"]
        for f in REQUIRED_CONTENU:
            _check(f in c and c[f], errors, f"contenu.{f} requis et non vide")
        if "gravite" in c:
            _check(c["gravite"] in VALID_GRAVITES, errors,
                   f"contenu.gravite invalide : '{c['gravite']}'. Acceptés : {VALID_GRAVITES}")
        if "type" in c:
            _check(c["type"] in VALID_TYPES, errors,
                   f"contenu.type invalide : '{c['type']}'. Acceptés : {VALID_TYPES}")
        if "exemple" not in c:
            warnings.append("contenu.exemple absent — fortement recommandé pour adoption")

    # ── Preuve
    if "preuve" in entry:
        p = entry["preuve"]
        _check("type" in p, errors, "preuve.type requis")
        if "type" in p:
            _check(p["type"] in VALID_PREUVES, errors,
                   f"preuve.type invalide : '{p['type']}'. Acceptés : {VALID_PREUVES}")
        if "reference" not in p or not p.get("reference"):
            warnings.append("preuve.reference absent — réduira la confiance des validateurs")

    # ── Validation / votes
    if "validation" in entry:
        v = entry["validation"]
        for f in REQUIRED_VALIDATION:
            _check(f in v, errors, f"validation.{f} requis")
        if "votes_requis" in v:
            vr = v["votes_requis"]
            _check(isinstance(vr, int) and vr >= 3, errors,
                   f"validation.votes_requis doit être ≥ 3 (actuel : {vr})")
        # Only check vote count when statut is 'actif'
        if entry.get("statut") == "actif" and "votes_requis" in v and "votes" in v:
            vr = v["votes_requis"]
            if len(v["votes"]) < vr:
                errors.append(f"Votes insuffisants pour statut 'actif': {len(v['votes'])}/{vr}")
        if "votes" in v:
            for i, vote in enumerate(v.get("votes", [])):
                _check("votant_did" in vote, errors, f"vote[{i}].votant_did manquant")
                _check("decision" in vote and vote.get("decision") in VALID_DECISIONS, errors,
                       f"vote[{i}].decision invalide. Acceptés : {VALID_DECISIONS}")
                _check("timestamp" in vote, errors, f"vote[{i}].timestamp manquant")
        # Positifs coherence
        if all(k in v for k in ["votes_positifs", "votes_total"]):
            _check(v["votes_positifs"] <= v["votes_total"], errors,
                   "votes_positifs ne peut pas dépasser votes_total")

    # ── Confiance
    if "confiance" in entry:
        conf = entry["confiance"]
        if "score" in conf:
            s = conf["score"]
            _check(0 <= s <= 100, errors, f"confiance.score doit être entre 0 et 100 (actuel : {s})")
            if s < CONFIDENCE_THRESHOLD_LOW:
                warnings.append(f"confiance.score faible ({s}) — entrée non appliquée automatiquement")

    # ── Statut
    if "statut" in entry:
        _check(entry["statut"] in VALID_STATUTS, errors,
               f"statut invalide : '{entry['statut']}'. Acceptés : {VALID_STATUTS}")

    # ── Source
    if "source" in entry:
        s = entry["source"]
        if s.get("affichage") in VALID_AFFICHAGE:
            aff = s["affichage"]
            if aff == "pseudonyme":
                _check("pseudonyme" in s and s["pseudonyme"], errors,
                       "source.pseudonyme requis quand affichage='pseudonyme'")
            if aff == "anonyme":
                _check("preuve_zkp" in s and s["preuve_zkp"], errors,
                       "source.preuve_zkp requis quand affichage='anonyme'")
            if aff in ("public", "pseudonyme"):
                _check("did" in s, errors, "source.did requis pour affichage public ou pseudonyme")
        else:
            errors.append(f"source.affichage invalide. Acceptés : {VALID_AFFICHAGE}")

    # ── Visibilité
    if "visibilite" in entry:
        niv = entry["visibilite"].get("niveau")
        _check(niv in VALID_VISIBILITE, errors,
               f"visibilite.niveau invalide : '{niv}'. Acceptés : {VALID_VISIBILITE}")
        if niv == "organisation":
            _check(entry["visibilite"].get("organisation_id"), errors,
                   "visibilite.organisation_id requis quand niveau='organisation'")

    # ── Relations
    if "relations" in entry:
        r = entry["relations"]
        if "parent" not in r:
            warnings.append("relations.parent absent — hiérarchie du registre recommandée")

    # ── Cycle de vie
    if "cycle_de_vie" in entry:
        cdv = entry["cycle_de_vie"]
        _check("date_creation" in cdv, errors, "cycle_de_vie.date_creation requis")

    return {
        "valid":         len(errors) == 0,
        "entry_id":      entry.get("id", "inconnu"),
        "errors":        errors,
        "warnings":      warnings,
        "summary":       f"{len(errors)} erreur(s), {len(warnings)} avertissement(s)"
    }


def print_report(result: dict):
    GREEN  = "\033[92m"
    RED    = "\033[91m"
    YELLOW = "\033[93m"
    CYAN   = "\033[96m"
    RESET  = "\033[0m"
    BOLD   = "\033[1m"

    status = f"{GREEN}✅ VALIDE{RESET}" if result["valid"] else f"{RED}❌ INVALIDE{RESET}"
    print(f"\n{BOLD}SMCU Validator — {result['entry_id']}{RESET}")
    print(f"Status : {status}")
    print(f"Résumé : {result['summary']}\n")

    if result["errors"]:
        print(f"{RED}{BOLD}Erreurs:{RESET}")
        for e in result["errors"]:
            print(f"  {RED}✗{RESET} {e}")

    if result["warnings"]:
        print(f"\n{YELLOW}{BOLD}Avertissements:{RESET}")
        for w in result["warnings"]:
            print(f"  {YELLOW}⚠{RESET} {w}")

    print()


def main():
    if len(sys.argv) < 2:
        print("Usage: python validate.py <chemin_vers_entree.json>")
        print("Exemple: python validate.py examples/software-dev.json")
        sys.exit(1)

    path = sys.argv[1]

    if not os.path.exists(path):
        print(f"Erreur : fichier introuvable : '{path}'")
        sys.exit(1)

    with open(path, "r", encoding="utf-8") as f:
        try:
            entry = json.load(f)
        except json.JSONDecodeError as e:
            print(f"Erreur JSON : {e}")
            sys.exit(1)

    result = validate(entry)
    print_report(result)

    # Output JSON for piping
    if "--json" in sys.argv:
        print(json.dumps(result, ensure_ascii=False, indent=2))

    sys.exit(0 if result["valid"] else 1)


if __name__ == "__main__":
    main()
